src/python/pants/backend/jvm/tasks/jvm_compile/zinc/zinc_compile.py

```diff
@@ -491,6 +492,46 @@ def _compile_hermetic(self, jvm_options, ctx, classes_dir, jar_file,
     # TODO: This should probably return a ClasspathEntry rather than a Digest
     return res.output_directory_digest
 
+  def _compute_scratch_inputs(self, classes_dir, relpath_to_analysis, jar_file):
+    """
+    Compute for the scratch inputs for ExecuteProcessRequest.
+
+    :param classes_dir: relative path to classes dir from buildroot
+    :param relpath_to_analysis: relative path to zinc analysis file from buildroot
+    :param jar_file: relative path to z.jar from buildroot
+    :return: digest of merged analysis file and loose class files.
+    """
+    if not os.path.exists(relpath_to_analysis):
+      return EMPTY_DIRECTORY_DIGEST
+
+    def _get_analysis_snaoshot():
+      _analysis_snapshot, = self.context._scheduler.capture_snapshots([
+        PathGlobsAndRoot(
+          PathGlobs([relpath_to_analysis]),
+          get_buildroot(),
+        ),
+      ])
+      return _analysis_snapshot
+
+    def _get_classes_dir_snapshot():
+      if self.get_options().use_classpath_jars:
+        with zipfile.ZipFile(jar_file, 'r') as zip_ref:
+          zip_ref.extractall(classes_dir)
+
+      _classes_dir_snapshot, = self.context._scheduler.capture_snapshots([
+        PathGlobsAndRoot(
+          PathGlobs([classes_dir + '/**']),
+          get_buildroot(),
+        ),
+      ])
+      return _classes_dir_snapshot
+
+    analysis_snapshot = _get_analysis_snaoshot()
+    classes_dir_snapshot = _get_classes_dir_snapshot()
```

Comment:
Why do we need a classes dir? I thought Zinc could operate purely on jars? If this _is_ necessary, can you add a comment to the code explaining why? This seems pretty weird...